### PR TITLE
fixes  #22662 Procs with constructor pragma doesn't initialize object's fields

### LINKED ISSUE #22662: Procs with constructor pragma doesn't initialize object's fields
### Description

Procedures defined with constructor pragma doesn't initialize object's fields.
Example:
testcpp2.nim:
```nim
type
  NimClass* = object
    x: int          # Not initialized

proc makeNimClass(): NimClass {.constructor.} =
  discard

proc main =
  for i in 0 .. 1:
    var n = makeNimClass()
    echo n.x
    n.x = 123
    echo n.x

main()
```

### Nim Version

Nim Compiler Version 2.1.1 [Linux: amd64]
Compiled at 2023-09-05
Copyright (c) 2006-2023 by Andreas Rumpf

git hash: 8f7aedb3d1941650e6650c07aa4bcf6a7dee0b90
active boot switches: -d:release

### Current Output

```text
268435456
123
123
123
```

### Expected Output

```text
0
123
0
123
```

### Possible Solution

When generating C++ code, initialize fields in constructor or in struct definition:
```c++
struct tyObject_NimClass {
  NI x{};
  // or NI x = 0;
};

// Or
tyObject_NimClass::tyObject_NimClass() {
  x = 0;
}
```

### Additional Information

In Nim generated C++ code, field `x` is not initialized:
```C++
struct tyObject_NimClass__ZFCcrp08suVyruEm60R7mA {
	NI x;
N_LIB_PRIVATE N_NOCONV(, tyObject_NimClass__ZFCcrp08suVyruEm60R7mA)(void);
};

...

N_LIB_PRIVATE N_NOCONV(, tyObject_NimClass__ZFCcrp08suVyruEm60R7mA::tyObject_NimClass__ZFCcrp08suVyruEm60R7mA)(void) {
	nimfr_("makeNimClass", "/home/player/worktmpfs/testnim/testcpp2.nim");
	popFrame();
}
```

When an object contains the struct type imported C++ code, it is initialized but Nim type fields are not.
```nim
{.emit:"""/*TYPESECTION*/
struct CppClass {
  int x;
};
""".}

type
  CppClass* {.importcpp.} = object
    x: cint

  NimClass* = object
    x: int          # Not initialized
    cpp: CppClass   # Initialized to zero

proc makeNimClass(): NimClass {.constructor.} =
  discard

proc main =
  for i in 0 .. 1:
    var n = makeNimClass()
    echo n.x, " ", n.cpp.x
    n.x = 123
    n.cpp.x = 321
    echo n.x, " ", n.cpp.x

main()
```

Output:
```
0 0
123 321
123 0
123 321
```

Expected output:
```
0 0
123 321
0 0
123 321
```

In Nim generated C++ code:
```c++
struct tyObject_NimClass__wPhP74Yg7WO49cvrXHdusjQ {
	NI x;
	CppClass cpp{};
N_LIB_PRIVATE N_NOCONV(, tyObject_NimClass__wPhP74Yg7WO49cvrXHdusjQ)(void);
};

...

N_LIB_PRIVATE N_NOCONV(, tyObject_NimClass__wPhP74Yg7WO49cvrXHdusjQ::tyObject_NimClass__wPhP74Yg7WO49cvrXHdusjQ)(void) {
	nimfr_("makeNimClass", "/home/player/worktmpfs/testnim/testcpp.nim");
	popFrame();
}
```

Field `cpp` is initialized as it has '{}' but field `x` is not initialized.

### PR DESCRIPTION
**Pull request by jmgomez**
fixes  #22662 Procs with constructor pragma doesn't initialize object's fields

### CODE REVIEW

**Comment by demotomohiro** [comment]
Thank you for fixing the bug!
I test this PR and fields are initialized properly.
But default values for object fields doesn't work.
https://nim-lang.github.io/Nim/manual.html#types-default-values-for-object-fields

For example:
```nim
type
  NimClass* = object
    x: int
    y = 777

proc makeNimClass(): NimClass {.constructor.} =
  discard

proc main =
  for i in 0 .. 1:
    var n = makeNimClass()
    echo n.x
    echo n.y
    n.x = 123
    echo n.x

main()
```

Current output:
```
0
0
123
0
0
123
```

Expected output:
```
0
777
123
0
777
123
```

In C++, non-static data members can have default value like Nim.
https://en.cppreference.com/w/cpp/language/data_members#Member_initialization
So when an object's field have non-zero default value, default member initializer can be used for it.

**Comment by jmgomez** [comment]
Thank you for looking it! 
But it seems like a different issue. For instance the following produces the same output:

```nim
type
  NimClass* = object
    x: int
    y = 777

proc createNimClas(): NimClass =
  discard

proc main =
  for i in 0 .. 1:
    var n = createNimClas()
    echo n.x
    echo n.y
    n.x = 123
    echo n.x

main()
```

Edit: Will try to see if I can manage to inject the Nim init inside the constructor anyways

File: compiler/ccgtypes.nim [hunk 1]
```diff
@@ -712,9 +719,9 @@ proc genRecordFieldsAux(m: BModule; n: PNode,
         result.addf("\t$1$4 $2:$3;$n", [getTypeDescAux(m, field.loc.t, check, dkField), sname, rope($field.bitsize), noAlias])
       else:
         # don't use fieldType here because we need the
-        # tyGenericInst for C++ template support
-        if fieldType.isOrHasImportedCppType():
-          result.addf("\t$1$3 $2{};$n", [getTypeDescAux(m, field.loc.t, check, dkField), sname, noAlias])
+        # tyGenericInst for C++ template support
```
Comment (Araq) [inline compiler/ccgtypes.nim]:
```suggestion
        # tyGenericInst for C++ template support
```

File: compiler/ccgtypes.nim [hunk 2]
```diff
@@ -713,8 +720,8 @@ proc genRecordFieldsAux(m: BModule; n: PNode,
       else:
         # don't use fieldType here because we need the
         # tyGenericInst for C++ template support
-        if fieldType.isOrHasImportedCppType():
-          result.addf("\t$1$3 $2{};$n", [getTypeDescAux(m, field.loc.t, check, dkField), sname, noAlias])
+        if fieldType.isOrHasImportedCppType() or hasCppCtor(m, field.owner.typ):
+          result.addf("\t$1$3 $2{};$n ", [getTypeDescAux(m, field.loc.t, check, dkField), sname, noAlias])
```
Comment (Araq) [inline compiler/ccgtypes.nim]:
```suggestion
          result.addf("\t$1$3 $2{};$n", [getTypeDescAux(m, field.loc.t, check, dkField), sname, noAlias])
```

**Pull request merged**